Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs

```diff
@@ -458,11 +459,172 @@ internal static bool Approximately(float a, in float b)
             }
         }
 
+        internal static TransformQvvs GetRootMotionDelta(ref MecanimControllerBlob controllerBlob,
+                                             ref SkeletonClipSetBlob clipSet,
+                                             NativeArray<MecanimParameter> parameters,
+                                             float deltaTime,
+                                             NativeList<TimedMecanimClipInfo> clipWeights,
+                                             DynamicBuffer<TimedMecanimClipInfo> previousFrameClipInfo,
+                                             float totalWeight,
+                                             float weightCullingThreshold)
+        {
+            //Get the current clip deltas
+            var currentRoot = TransformQvvs.identity;
+            for (int i = 0; i < clipWeights.Length; i++)
+            {
+                var clipWeight = clipWeights[i];
+                ref var clip = ref clipSet.clips[clipWeight.mecanimClipIndex];
+                var blendWeight = clipWeight.weight / totalWeight;
+                //Cull clips with negligible weight
+                if (blendWeight < weightCullingThreshold)
+                    continue;
+
+                ref var state = ref controllerBlob.layers[clipWeight.layerIndex].states[clipWeight.stateIndex];
+                var time = state.isLooping ? clip.LoopToClipTime(clipWeight.motionTime) : math.min(clipWeight.motionTime, clip.duration);
+                var stateSpeed = state.speedMultiplierParameterIndex != -1 ?
+                                 parameters[state.speedMultiplierParameterIndex].floatParam * state.speed :
+                                 state.speed;
+                var speedModifiedDeltaTime = deltaTime * stateSpeed;
+                var deltaTransform = TransformQvvs.identity;
+                var hasLooped = state.isLooping && time - deltaTime < 0f;
+
+                //If the clip has looped, get a sample of the end of the clip to incorporate it into the delta
+                if (hasLooped)
+                {
+                    deltaTransform = clip.SampleBone(0, time);
+                    var previousClipSample = clip.SampleBone(0, time - speedModifiedDeltaTime);
+
+                    var sampleEnd = math.select(clip.duration, -clip.duration, stateSpeed < 0f);
+                    var endClipSample = clip.SampleBone(0, sampleEnd);
+
+                    deltaTransform.position += endClipSample.position - previousClipSample.position;
```

Comment:
I know you are sick of me discussing this, and I am willing to let this through as you have spent more time testing this, but I don't think the math is correct here.

Imagine we have an animation for 1 second duration that moves the object on the x axis from 0 to 1. Now imagine that our speedModifiedDeltaTime is 0.01 and is equivalent to deltaTime and our current time in the clip is 0.005 which means we looped. Naturally, the expected root motion delta should also be 0.01.

We sample deltaTransform and get a value of 0.005. For previousClipSample, we sample at -0.005 which ACL clamps to 0 (ACL clamps all time values from 0 to the clip duration), so we get a value of 0 there. And for endClipSample, we sample at clip duration and get the value of 1.

Thus our final delta = 0.005 + 1 - 0 = 1.005 which does not equal 0.01.